### ISSUE TITLE
News Downloader doesn't download sample Reuters feeds (full download)

### ISSUE THREAD

**Issue by mwoz123** [koreader/koreader#3596]
* KOReader version: ~ git-1411
* Device: Inkbook 

#### Issue
Full download article doesn't work for sample config with Reuters, but changed (from download description) to full article.
Returned 301 http status ("301 Moved Permanently") not handled.

Here I think we should ask a question - would we like to handle it manually, manually code this issue (might happen that we will find more, unhandled statuses in feature, that won't work) or replace it with qualified backed (https://github.com/koreader/koreader/issues/3516) ?

I probably won't be able to code handling for this status.

#### Steps to reproduce
Edit sample config, change Reuters to: `download_full_article=true`; download news.

**Comment by Frenzie** [comment]
Ah yes, that looks like it involves HTTPS and only HTTP redirects are currently handled. 

Like I said in the other issue [lua-http](https://github.com/daurnimator/lua-http) looks like a very interesting option. Unfortunately it's one of those things that's not necessarily difficult, but that I do expect do be somewhat time-consuming to implement.

**Comment by mwoz123** [comment]
I did some draft work to change backend engine (#3618)

I tried to create library for lua-http but I got stuck on my Ubuntu 16.04 (only available luarocks 2.2.0) with :

>Segmentation fault (core dumped)
*** Segmentation fault

@Frenzie could you create a library from lua-http, so it can be used in project , and add it to my (#3618) ?
Thanks

**Comment by Frenzie** [comment]
Unfortunately that's a bit of a long-term goal. I'm not sure I have the time for it at present.

**Comment by Frenzie** [comment]
Btw, it's also a semi-blocker on https://github.com/koreader/koreader/pull/3176 (Or rather, for the moment I'm blocking myself from just stripping out a few dics that otherwise wouldn't work.)

**Comment by poire-z** [comment]
Our current http module is quite low-level, but it allows you to do a few things (example in [wikipedia.lua](https://github.com/koreader/koreader/blob/e928276d731cb9bb924b4f747a271d7178216e7e/frontend/ui/wikipedia.lua#L119)).
If you'd just wanted to handle 301 and 302 redirect, you could use something like
```lua
-- https://en.wikipedia.org/wiki/List_of_HTTP_status_codes#3xx_Redirection
if code and string.sub(code, 1, 1) ~= "3" then -- handle 301, 302...
  if headers and headers["location"] then
    local redirected_url = headers["location"]
    -- clean up stuff, and recursively recall your download function
    -- (you may want to add a counter, so you're not in an infinite loop
    -- if the remote server is misconfigured and keeps redirecting)
    return your_download_function(redirected_url)
  end
end
```

**Comment by Frenzie** [comment]
> If you'd just wanted to handle 301 and 302 redirect, you could use something like

Not for GH which explicitly depends on redirection within the same SSL connection, but it might work for Reuters.

**Comment by Frenzie** [comment]
Also see https://github.com/koreader/koreader/commit/77875852af3ce2deaefb136aa7d1673bfef243ae

**Comment by Frenzie** [comment]
PS Back then (three months ago) I also tried lua-http and that Just Works™.

PPS HTTP redirects just fine. It's Luasec that doesn't handle redirects. Just to be clear. ;-)

**Comment by poire-z** [comment]
I see. It feels strange they can impose it to be within the same SSL connection.
If it's just the cookie you're trying to handle that they expect to find, may be you have a bug: in the _request_ headers, the header is simply named `Cookie:` (https://stackoverflow.com/questions/3467114/how-are-cookies-passed-in-the-http-protocol), and not `set-cookie` like it looks you have it.

**Comment by Frenzie** [comment]
Iirc it's not just the cookie and as soon as it became clear that it'd require more reverse-engineering of whatever GH does at the risk of it breaking tomorrow, combined with a quick test showing that the much more versatile lua-http Just Works™, in my mind it was settled that implementing lua-http would be much  more fruitful.

Anyway, for the dictionary thing it's obviously not an actual blocker. It's just an (perhaps obviously failing) attempt at motivating myself to deal with potential Android complications (lpeg is known; there might be more).

**Comment by mwoz123** [comment]
>PS Back then (three months ago) I also tried lua-http and that Just Works™.

@Frenzie do you still have it on your drive? could you send me e.g. via email?
I'm not able to build/compile it with dependencies:(

**Comment by Frenzie** [comment]
> I tried to create library for lua-http but I got stuck on my Ubuntu 16.04 (only available luarocks 2.2.0) with :

By testing in this context I mean using either `dofile` or alternatively typing or copying choice commands in the Lua prompt (don't use locals). That's the advantage of scripting languages like Lua — they don't need to be compiled. ;-)

You can also use compiled luarocks objects in KOReader's LuaJIT for testing things in the full UX if you want, of course. But lua-http can be used as a drop-in replacement, so really there's nothing that needs to be done in that regard.

That being said, presumably the main problem is that you were trying to use it in combination with Lua 5.3, while LuaJIT is 5.1 (and to some degree 5.2) ABI compatible. The luarocks version shouldn't be relevant.
